modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/IdempotentRequestHandler.java

```diff
@@ -0,0 +1,144 @@
+    public static final String STATUS_PREFIX = "s_";
+
+    /** @since 11.5 */
+    public static final String CONTENT_PREFIX = "c_";
+
+    protected final String store;
+
+    protected final int ttl;
+
+    public IdempotentRequestHandler(FilterConfig filterConfig) {
+        // default values
+        int ttl = 330;
+        String store = "requestcontroller";
+        // retrieve potential configuration from filter
+        String ttlConfig = filterConfig.getInitParameter(IdempotentRequestHandler.TTL_SECONDS_PROPERTY);
+        if (ttlConfig != null) {
+            try {
+                ttl = Integer.valueOf(ttlConfig);
+            } catch (NumberFormatException e) {
+                log.error(e, e);
+            }
+        }
+        String storeConfig = filterConfig.getInitParameter(IdempotentRequestHandler.STORE_PROPERTY);
+        if (storeConfig != null) {
+            store = storeConfig;
+        }
+        this.store = store;
+        this.ttl = ttl;
+    }
+
+    public IdempotentRequestHandler(String storename, int ttl) {
+        this.store = storename;
+        this.ttl = ttl;
+    }
+
+    public void handle(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        if (!"POST".equals(request.getMethod())) {
+            chain.doFilter(request, response);
+            return;
+        }
+        String key = request.getHeader(IdempotentRequestHandler.HEADER_KEY);
+        if (key == null) {
+            chain.doFilter(request, response);
+            return;
+        }
+        KeyValueStore store = Framework.getService(KeyValueService.class).getKeyValueStore(this.store);
+        String storedContent = store.getString(IdempotentRequestHandler.CONTENT_PREFIX + key);
+        if (storedContent == null) {
+            // new request
+            boolean cleanup = false;
+            store.put(IdempotentRequestHandler.CONTENT_PREFIX + key, IdempotentRequestHandler.INPROGRESS_MARKER, ttl);
+            try {
+                CopyingResponseWrapper wrapper = new CopyingResponseWrapper(response);
+                chain.doFilter(request, wrapper);
+                store.put(IdempotentRequestHandler.CONTENT_PREFIX + key, wrapper.getCaptureAsString(), ttl);
+                store.put(IdempotentRequestHandler.STATUS_PREFIX + key, String.valueOf(wrapper.getStatus()), ttl);
+            } catch (IOException | ServletException e) {
+                cleanup = true;
+                throw e;
+            } finally {
+                if (cleanup || response.getStatus() >= 400) {
+                    // error request: cleanup store
+                    store.put(IdempotentRequestHandler.CONTENT_PREFIX + key, (String) null, ttl);
+                    store.put(IdempotentRequestHandler.STATUS_PREFIX + key, (String) null, ttl);
+                }
+            }
+        } else if (IdempotentRequestHandler.INPROGRESS_MARKER.equals(storedContent)) {
```

Comment:
For in-progress it's better to use a special status value (negative?) instead of the content. This way the content is pure content and not mixed with other logic.

Reply:
I kept this key as is in place of json serialization of response headers + status, let me know if you still think I should change this marker (sorry, renamed and moved all classes...)